### ISSUE TITLE
Qt6 macdeployqt not copying plugins

### ISSUE THREAD

**Issue by csparker247** [Homebrew/homebrew-core#73019]
### `brew gist-logs <formula>` link OR `brew config` AND `brew doctor` output

brew config:
HOMEBREW_VERSION: 3.0.5
ORIGIN: https://github.com/Homebrew/brew
HEAD: 8b677c0ddd5f96b5521c1b353a9ca07dfe57b80e
Last commit: 3 days ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 0d5ae88abf46a18896a1fd0b481a6cfe2e28e0ff
Core tap last commit: 2 days ago
Core tap branch: master
HOMEBREW_PREFIX: /usr/local
HOMEBREW_CASK_OPTS: []
HOMEBREW_MAKE_JOBS: 4
Homebrew Ruby: 2.6.3 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: quad-core 64-bit broadwell
Clang: 12.0 build 1200
Git: 2.30.2 => /usr/local/bin/git
Curl: 7.64.1 => /usr/bin/curl
macOS: 11.2.1-x86_64
CLT: 12.4.0.0.1.1610135815
Xcode: N/A

brew doctor:
Your system is ready to brew.

### 

- [X] The `brew doctor` above contains no "Warning" lines.

### What were you trying to do (and why)?

Trying to package a CMake-configured Qt6 application bundle using macdeployqt.

### What happened (include all command output)?

Executable builds and runs correctly, but without deployed libraries in bundle.`macdeployqt` runs on app bundle without warnings, but verbose logging indicates it is looking for plugins in the wrong directory. Post-deployed bundle throws assert because plugin can't be found. Bundle does not contain the `PlugIns` directory or linked plugins.

This same process works correctly with `qt@5`. My understanding is that this is an installation configuration error and not a Qt bug proper. See [this old issue](https://github.com/cartr/homebrew-qt4/issues/38) for something similar. If that's incorrect, I can post this is a bug on the Qt bug tracker.

The following is an abbreviated output from running `macdeployqt` on the test bundle. The removed portions show it working as expecting, detecting Qt frameworks and copying linked libraries into the bundle:
```bash 
$ ./test.app/Contents/MacOS/test
Success.
$ macdeployqt test.app -verbose=3
Log: Using otool:
Log:  inspecting "[REDACTED]/qt6-test/build/test.app/Contents/MacOS/test"
Log: Using otool:
Log:  inspecting "/usr/local/opt/qt/lib/QtSql.framework/Versions/A/QtSql"
Log: Adding framework:
Log: Framework name "QtSql.framework" 
 Framework directory "/usr/local/opt/qt/lib/" 
 Framework path "/usr/local/opt/qt/lib/QtSql.framework" 
 Binary directory "Versions/A" 
 Binary name "QtSql" 
 Binary path "/Versions/A/QtSql" 
 Version "A" 
 Install name "/usr/local/opt/qt/lib/QtSql.framework/Versions/A/QtSql" 
 Deployed install name "@executable_path/../Frameworks/QtSql.framework/Versions/A/QtSql" 
 Source file Path "/usr/local/opt/qt/lib/QtSql.framework/Versions/A/QtSql" 
 Framework Destination Directory (relative to bundle) "Contents/Frameworks/QtSql.framework" 
 Binary Destination Directory (relative to bundle) "Contents/Frameworks/QtSql.framework/Versions/A" 
...
Log: Deploying plugins from "/usr/local/plugins"
Log: Created configuration file: "test.app/Contents/Resources/qt.conf"
Log: This file sets the plugin search path to "test.app/Contents/PlugIns"
Log: Using strip:
Log:  stripped "[REDACTED]/qt6-test/build/test.app/Contents/MacOS/test"
$ ./test.app/Contents/MacOS/test 
ASSERT: "QSqlDatabase::isDriverAvailable("QSQLITE")" in file /Users/seth/source/qt6-test/main.cpp, line 8
zsh: abort      ./test.app/Contents/MacOS/test
```

### What did you expect to happen?

App is bundled correctly and executes.

### Step-by-step reproduction instructions (by running `brew` commands)

Put the following two files in a directory.

CMakeLists.txt:
```cmake
cmake_minimum_required(VERSION 3.19.0)
project(test VERSION 1.0.0 LANGUAGES CXX)
set(CMAKE_CXX_STANDARD 17)
set(CMAKE_CXX_STANDARD_REQUIRED ON)
set(CMAKE_AUTOMOC ON)
set(CMAKE_AUTORCC ON)
set(CMAKE_AUTOUIC ON)
find_package(Qt6 COMPONENTS Core Sql REQUIRED)
add_executable(test MACOSX_BUNDLE main.cpp)
target_link_libraries(test PRIVATE Qt6::Core Qt6::Sql)
```

main.cpp:
```cpp
#undef QT_NO_DEBUG
#include <QCoreApplication>
#include <QtSql>
#include <QDebug>
int main(int argc, char *argv[])
{
  QCoreApplication a(argc, argv);
  Q_ASSERT(QSqlDatabase::isDriverAvailable("QSQLITE"));
  qDebug() << "Success.";
  return 0;
}
```

Run these commands:
```shell
brew install cmake qt

# Build app
mkdir build && cd build
cmake ..
make

# Test that pre-deploy executable works
./test.app/Contents/MacOS/test # prints Success.

# Deploy app bundle
macdeployqt test.app -verbose=3 

# Test that post-deploy executable works  
./test.app/Contents/MacOS/test # Hits the ASSERT

**Comment by carlocab** [comment]
Does installing one of the SQL database drivers help? i.e. one of
```
qt-percona-server
qt-mysql
qt-libiodbc
qt-mariadb
qt-postgresql
qt-unixodbc
```

**Comment by csparker247** [comment]
Testing with `qt-mysql`, it does not. This is actually just a toy example. My full app also uses QWidgets and thus requires the `qcocoa` plugin as well, but that is also missing from the bundle. This issue appears to affect all the deployment of all plugins.

As a side note, I've tried the `macdeployqt -libpath=/usr/local/share/qt/` but that also doesn't help. I think that only helps resolve linked libs and not plugins.

**Comment by carlocab** [comment]
@paperchalice, could you take a look at this?

**Comment by paperchalice** [comment]
Reported, but no replies now: https://bugreports.qt.io/browse/QTBUG-91644

**Comment by carlocab** [comment]
I'm unsure as to why you're hitting the `ASSERT`. The formula `test` block involves an extended version of your reproduction steps:

https://github.com/Homebrew/homebrew-core/blob/77474e6293025aeea3a7e25e86883967eaa02675/Formula/qt.rb#L163-L193

And yet it seems to have run fine in CI?

**Comment by paperchalice** [comment]
mentioned here: https://github.com/Homebrew/homebrew-core/pull/70826#issuecomment-789352857

**Comment by paperchalice** [comment]
A temporary solution is change [this line](https://github.com/qt/qttools/blob/bf9f5e2f70ef676cbc5a96aa2dbb82e13ca2563d/src/macdeployqt/macdeployqt/main.cpp#L241) to `deploymentInfo.pluginPath = deploymentInfo.qtPath + "/share/qt/plugins";`

**Comment by csparker247** [comment]
> I'm unsure as to why you're hitting the `ASSERT`. The formula `test` block involves an extended version of your reproduction steps:
> 
> 
> And yet it seems to have run fine in CI?

Yeah, I basically made my toy example from the CI code so it would be proven and testable. Mine does run correctly _before_ I process the bundle with `macdeployqt`. The CI script only tests an executable linked against the system-installed libs. After you run `macdeployqt`, it updates the linkage to point to a package-installed library relative to the executable. Apparently `macdeployqt` doesn't check that it actually found and copied the plugins successfully but updates the executable anyway? 🤷 

> mentioned here: [#70826 (comment)](https://github.com/Homebrew/homebrew-core/pull/70826#issuecomment-789352857)
> Because `macdeployqt` still search plugins in `HOMEBREW_PREFIX/plugins` not in the `share/qt/plugins`.

Ah ha! So it is a Qt bug. That makes sense. In the short term, I can still build against the Qt5 libs for awhile. In the more medium term, I think making my own symlink at `/usr/local/plugins/` is the easiest workaround, but I'll keep the other patch in mind in case this persists for longer.

As far as I'm concerned, you can close this issue, but I understand if you want to keep it open to keep tracking the upstream issue. Thanks for all of the help!

**Comment by carlocab** [comment]
>I think making my own symlink at `/usr/local/plugins/` is the easiest workaround

I wouldn't be opposed to `brew` writing the appropriate symlinks to `HOMEBREW_PREFIX/plugins` for the formula too, but it would be good to try to get this resolved upstream first.

**Comment by jonaski** [comment]
Unfortunately creating a symlink isn't enough. macdeployqt isn't doing it's job on the plugins. install_name_tool does not look to be run at all, at least the sqlite and jpeg plugin has incorrect paths. Check it with otool -L on the plugins.

**Comment by jonaski** [comment]
```
Log:  inspecting "strawberry.app/Contents/PlugIns/imageformats/libqjpeg.dylib"
ERROR: Cannot resolve rpath "@rpath/QtGui.framework/Versions/A/QtGui"
ERROR:  using QSet("/Users/jonas/build/strawberry/build/lib")
ERROR: Cannot resolve rpath "@rpath/QtCore.framework/Versions/A/QtCore"
ERROR:  using QSet("/Users/jonas/build/strawberry/build/lib")
```

Full log:

[macdeployqt.log](https://github.com/Homebrew/homebrew-core/files/6152905/macdeployqt.log)

(modified macdeployqt to specify -plugins-dir parameter and dump the install_name_tool commadn in runInstallNameTool().

**Comment by jonaski** [comment]
The problem is that the second line from otool for the Qt plugins is not the plugin itself, but a dependency, in the case for libqjpeg.dylib, this is libjpeg.9.dylib.
This causes libjpeg not to be added to info.dependencies in findDependencyInfo()

A dirty fix I did was to remove the second outputLines.removeFirst(); line in findDependencyInfo().
This fixed the plugin deployment, but causes some extra loops. So the patch is not perfect, but at least it works for me for now. I never got time to look more into it as I'm having some health issues too.

Patch:


```
diff -ruN a/src/macdeployqt/shared/shared.cpp b/src/macdeployqt/shared/shared.cpp
--- a/src/macdeployqt/shared/shared.cpp	2021-02-25 03:26:59.000000000 +0100
+++ b/src/macdeployqt/shared/shared.cpp	2021-03-17 21:40:49.505618152 +0100
@@ -201,7 +201,6 @@
         } else {
             LogError() << "Could not parse otool output line:" << outputLines.first();
         }
-        outputLines.removeFirst();
     }
 
     for (const QString &outputLine : outputLines) {

```

otool output for libqjpeg.dylib:


```
/usr/local//Cellar/qt/6.0.2/share/qt/plugins/imageformats/libqjpeg.dylib:
	/usr/local/opt/jpeg/lib/libjpeg.9.dylib (compatibility version 14.0.0, current version 14.0.0)
	@rpath/QtGui.framework/Versions/A/QtGui (compatibility version 6.0.0, current version 6.0.2)
	/System/Library/Frameworks/OpenGL.framework/Versions/A/OpenGL (compatibility version 1.0.0, current version 1.0.0)
	/System/Library/Frameworks/AGL.framework/Versions/A/AGL (compatibility version 1.0.0, current version 1.0.0)
	/System/Library/Frameworks/AppKit.framework/Versions/C/AppKit (compatibility version 45.0.0, current version 2022.20.117)
	/System/Library/Frameworks/ImageIO.framework/Versions/A/ImageIO (compatibility version 1.0.0, current version 1.0.0)
	/System/Library/Frameworks/Metal.framework/Versions/A/Metal (compatibility version 1.0.0, current version 244.32.7)
	@rpath/QtCore.framework/Versions/A/QtCore (compatibility version 6.0.0, current version 6.0.2)
	/System/Library/Frameworks/DiskArbitration.framework/Versions/A/DiskArbitration (compatibility version 1.0.0, current version 1.0.0)
	/System/Library/Frameworks/IOKit.framework/Versions/A/IOKit (compatibility version 1.0.0, current version 275.0.0)
	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 904.4.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1292.60.1)

```

There is a modified version here in case anyone is interested: 
https://github.com/strawberrymusicplayer/strawberry/tree/master/3rdparty/macdeployqt
It also fixes the plugin dir issue and deploys gstreamer plugins.

**Comment by csparker247** [comment]
I can confirm that @jonaski's version of `macdeployqt` does properly copy the missing plugins for my use case. A good replacement in lieu of the upstream fix being deployed. Here's a standalone CMakeLists.txt for building it:

```cmake
project(macdeployqt)
cmake_minimum_required(VERSION 3.11)

find_package(Qt6 REQUIRED COMPONENTS Core)

add_executable(macdeployqt main.cpp shared.cpp)
target_link_libraries(macdeployqt PRIVATE
  "-framework AppKit"
  Qt6::Core
)
```

**Comment by csparker247** [comment]
A patch request is in review for this upstream. This should be resolved soon.

**Comment by csparker247** [comment]
This should now be resolved upstream. The Qt bug tracker has it targeting Qt 6.2.0 Alpha, but since that's not slated until September, it probably will make it into a release before then. Possibly even the 6.1 release in early May.

**Comment by carlocab** [comment]
This can also be fixed for the current version of the formula if you have a patch you can apply to it.

**Comment by csparker247** [comment]
The fix is in qttools, which is a git submodule inside of the qt super project. The below file is the git patch relative to the qttools dir. I'm not sure how to apply that to the "collapsed" source tree for the qt super project, but I'll leave it here in case anyone else already knows how.

[macdeployqt_fix-deploy-plugin.txt](https://github.com/Homebrew/homebrew-core/files/6387443/macdeployqt_fix-deploy-plugin.txt)

**Comment by csparker247** [comment]
These fixes have been cherry picked for 6.1.
